src/unix/linux.c

```diff
@@ -298,6 +593,244 @@ int uv__io_check_fd(uv_loop_t* loop, int fd) {
+  sqe->len = 0xFFF; /* STATX_BASIC_STATS + STATX_BTIME */
+  sqe->opcode = UV__IORING_OP_STATX;
+
+  if (is_fstat) {
+    sqe->addr = (uintptr_t) "";
+    sqe->fd = req->file;
+    sqe->statx_flags |= 0x1000; /* AT_EMPTY_PATH */
+  }
+
+  if (is_lstat)
+    sqe->statx_flags |= AT_SYMLINK_NOFOLLOW;
+
+  uv__iou_submit(iou);
+
+  return 1;
+}
+
+
+void uv__statx_to_stat(const struct uv__statx* statxbuf, uv_stat_t* buf) {
+  buf->st_dev = makedev(statxbuf->stx_dev_major, statxbuf->stx_dev_minor);
+  buf->st_mode = statxbuf->stx_mode;
+  buf->st_nlink = statxbuf->stx_nlink;
+  buf->st_uid = statxbuf->stx_uid;
+  buf->st_gid = statxbuf->stx_gid;
+  buf->st_rdev = makedev(statxbuf->stx_rdev_major, statxbuf->stx_rdev_minor);
+  buf->st_ino = statxbuf->stx_ino;
+  buf->st_size = statxbuf->stx_size;
+  buf->st_blksize = statxbuf->stx_blksize;
+  buf->st_blocks = statxbuf->stx_blocks;
+  buf->st_atim.tv_sec = statxbuf->stx_atime.tv_sec;
+  buf->st_atim.tv_nsec = statxbuf->stx_atime.tv_nsec;
+  buf->st_mtim.tv_sec = statxbuf->stx_mtime.tv_sec;
+  buf->st_mtim.tv_nsec = statxbuf->stx_mtime.tv_nsec;
+  buf->st_ctim.tv_sec = statxbuf->stx_ctime.tv_sec;
+  buf->st_ctim.tv_nsec = statxbuf->stx_ctime.tv_nsec;
+  buf->st_birthtim.tv_sec = statxbuf->stx_btime.tv_sec;
+  buf->st_birthtim.tv_nsec = statxbuf->stx_btime.tv_nsec;
+  buf->st_flags = 0;
+  buf->st_gen = 0;
+}
+
+
+static void uv__iou_fs_statx_post(uv_fs_t* req) {
+  struct uv__statx* statxbuf;
+  uv_stat_t* buf;
+
+  buf = &req->statbuf;
+  statxbuf = req->ptr;
+  req->ptr = NULL;
+
+  if (req->result == 0) {
+    uv__statx_to_stat(statxbuf, buf);
+    req->ptr = buf;
+  }
+
+  uv__free(statxbuf);
+}
+
+
+static void uv__poll_io_uring(uv_loop_t* loop,
+                              struct uv__iou* iou,
+                              int ringfd,
+                              int events) {
+  struct uv__io_uring_cqe* cqe;
+  struct uv__io_uring_cqe* e;
+  uv_fs_t* req;
+  uint32_t head;
+  uint32_t tail;
+  uint32_t mask;
+  uint32_t i;
+
+  head = *iou->cqhead;
+  tail = atomic_load_explicit((_Atomic uint32_t*) iou->cqtail,
+                              memory_order_acquire);
+  mask = iou->cqmask;
+  cqe = iou->cqe;
+
+  for (i = head; i != tail; i++) {
+    e = &cqe[i & mask];
+
+    req = (uv_fs_t*) (uintptr_t) e->user_data;
+    assert(req->type == UV_FS);
+
+    uv__req_unregister(loop, req);
+    iou->in_flight--;
+
+    /* io_uring stores error codes as negative numbers, same as libuv. */
+    req->result = e->res;
+
+    switch (req->fs_type) {
+      case UV_FS_FSTAT:
+      case UV_FS_LSTAT:
+      case UV_FS_STAT:
+        uv__iou_fs_statx_post(req);
+        break;
+      default:  /* Squelch -Wswitch warnings. */
+        break;
+    }
+
+    uv__metrics_update_idle_time(loop);
+    req->cb(req);
+  }
+
+  atomic_store_explicit((_Atomic uint32_t*) iou->cqhead,
+                        tail,
+                        memory_order_release);
+
+  uv__metrics_inc_events(loop, 1);
```

Comment:
Not 100% sure, but shouldn't this be increased by the number of completions just handled?

Reply:
metrics_info_check starts failing when I do that, cc @trevnorris.

Reply:
My initial thought is to treat uring events like normal events. Do that by incrementing `nevents`. That should also take care of returning early so there's no need for the additional `has_iou_events` variable. This diff passes all tests locally, but I'm going to write an additional test right now to double check handling of uring events:
```diff
diff --git a/src/unix/linux.c b/src/unix/linux.c
index d9eb9fec..e5f88ca1 100644
--- a/src/unix/linux.c
+++ b/src/unix/linux.c
@@ -782,6 +782,6 @@ static void uv__iou_fs_statx_post(uv_fs_t* req) {
 
-static void uv__poll_io_uring(uv_loop_t* loop,
-                              struct uv__iou* iou,
-                              int ringfd,
-                              int events) {
+static int uv__poll_io_uring(uv_loop_t* loop,
+                             struct uv__iou* iou,
+                             int ringfd,
+                             int events) {
   struct uv__io_uring_cqe* cqe;
@@ -793,2 +793,3 @@ static void uv__poll_io_uring(uv_loop_t* loop,
   uint32_t i;
+  uint32_t completed;
 
@@ -799,2 +800,3 @@ static void uv__poll_io_uring(uv_loop_t* loop,
   cqe = iou->cqe;
+  completed = 0;
 
@@ -824,2 +826,3 @@ static void uv__poll_io_uring(uv_loop_t* loop,
     req->cb(req);
+    completed++;
   }
@@ -830,3 +833,3 @@ static void uv__poll_io_uring(uv_loop_t* loop,
 
-  uv__metrics_inc_events(loop, 1);
+  return completed;
 }
@@ -855,3 +858,2 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
   uint64_t base;
-  int have_iou_events;
   int have_signals;
@@ -995,3 +997,2 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
 
-    have_iou_events = 0;
     have_signals = 0;
@@ -1021,4 +1022,3 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
       if (fd == iou->ringfd) {
-        uv__poll_io_uring(loop, iou, fd, pe->events);
-        have_iou_events = 1;
+        nevents += uv__poll_io_uring(loop, iou, fd, pe->events);
         continue;
@@ -1097,5 +1097,2 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
 
-    if (have_iou_events != 0)
-      return;  /* Event loop should cycle now so don't poll again. */
-
     if (have_signals != 0)
```

Reply:
Just finished writing my first test, and appears that the way metrics are currently collected for these fs operations is broken. It's only incrementing by 1 for these all these fs events per loop instead of by the total number of fs events processed in the loop.

So, I have some debugging to do and more tests to write to see the extent of the problem. Not sure yet if this is specific to fs, or if it's a thread pool thing.

Here's the test I wrote that currently fails on `v1.x` but passes with the patch above in case you want to add it. Either way I'll be opening a PR to fix this: https://gist.github.com/trevnorris/bdaaf9e01cad30e1e509981fe0a8a8dd

Reply:
@bnoordhuis in short, I need to write a PR to fix the metrics counter anyways. So you can choose to incorporate the patch I posted above or I'll be adding it as part of my fix.